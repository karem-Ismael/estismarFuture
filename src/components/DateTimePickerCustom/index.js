export { default as DateTimePickerCustom } from "./DateTimePickerCustom";
