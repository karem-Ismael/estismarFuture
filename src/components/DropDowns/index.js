export { CitiesDropDown } from "./CitiesDropDown";
export { MakesDropDown } from "./MakesDropDown";
export { StatusDropDown } from "./StatusDropDown";
