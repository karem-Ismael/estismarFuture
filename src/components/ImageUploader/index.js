export { default as FileUploader } from "./ImageUploader";
