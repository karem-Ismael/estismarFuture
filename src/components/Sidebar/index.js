/* eslint-disable prettier/prettier */
/* eslint-disable react/no-unused-state */
/* eslint-disable prettier/prettier */
/**
 * Reactify Sidebar
 */
import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import { withRouter, Link } from "react-router-dom";
import { connect } from "react-redux";

import { Scrollbars } from "react-custom-scrollbars";

// redux actions
import { collapsedSidebarAction } from "Actions";
import jwt_decode from "jwt-decode";

// components
import UserBlock from "./UserBlock";
import SidebarContent from "./SidebarContent";
import AgencySidebar from "../AgencyMenu/AgencySidebar";
import UserProfile from "./Profile";
import LogoImage from "../../assets/img/img-logo.png"

class Sidebar extends Component {
  UNSAFE_componentWillMount() {
    this.updateDimensions();
  }

  shouldComponentUpdate(nextProps) {
    const {
      enableSidebarBackgroundImage,
      selectedSidebarImage,
      isDarkSidenav,
      locale,
    } = this.props;
    if (
      enableSidebarBackgroundImage !== nextProps.enableSidebarBackgroundImage ||
      selectedSidebarImage !== nextProps.selectedSidebarImage ||
      isDarkSidenav !== nextProps.isDarkSidenav ||
      locale
    ) {
      return true;
    }
    return false;
  }

  componentDidMount() {
    window.addEventListener("resize", this.updateDimensions);
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.updateDimensions);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const { windowWidth } = this.state;
    const { collapsedSidebar } = this.props;
    if (nextProps.location !== this.props.location) {
      if (windowWidth <= 1199) {
        this.props.collapsedSidebarAction(false);
      }
    }
  }

  updateDimensions = () => {
    this.setState({ windowWidth: window.innerWidth, windowHeight: window.innerHeight });
  };

  render() {
    const {
      enableSidebarBackgroundImage,
      selectedSidebarImage,
      isDarkSidenav,
      agencySidebar,
    } = this.props;

    return (
      <>
        <div
          className={classNames("rct-sidebar")}
          style={{
            backgroundImage: enableSidebarBackgroundImage ? `url(${selectedSidebarImage})` : "none",
          }}
        >
          <div
            className={classNames("rct-sidebar-content", {
              "sidebar-overlay-dark": isDarkSidenav,
              "sidebar-overlay-light": !isDarkSidenav,
            })}
          >
            <div className="site-logo d-flex justify-content-around">
              <Link to="/" className="logo-mini">
                <img
                  src={LogoImage}
                  className="mt-2"
                  alt="site logo"
                  width="55"
                  height="55"
                />
              </Link>
            </div>
            {/* <UserProfile /> */}

            <div className="rct-sidebar-wrap">
              <Scrollbars
                className="rct-scroll"
                autoHide
                autoHideDuration={100}
                style={{ height: "calc(100vh - 60px)" }}
              >
                <SidebarContent />
              </Scrollbars>
            </div>
          </div>
        </div>
      </>
    );
  }
}

// map state to props
const mapStateToProps = ({ settings }) => {
  const {
    enableSidebarBackgroundImage,
    selectedSidebarImage,
    collapsedSidebar,
    isDarkSidenav,
    locale,
  } = settings;
  return {
    enableSidebarBackgroundImage,
    selectedSidebarImage,
    collapsedSidebar,
    isDarkSidenav,
    locale,
  };
};

Sidebar.propTypes = {
  enableSidebarBackgroundImage: PropTypes.bool,
  selectedSidebarImage: PropTypes.any,
  isDarkSidenav: PropTypes.bool,
  locale: PropTypes.object,
  collapsedSidebar: PropTypes.func,
  collapsedSidebarAction: PropTypes.func,
  location: PropTypes.object,
  agencySidebar: PropTypes.any,
};

export default withRouter(connect(mapStateToProps, { collapsedSidebarAction })(Sidebar));
