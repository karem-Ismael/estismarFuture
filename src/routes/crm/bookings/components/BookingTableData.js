/* eslint-disable prettier/prettier */
import React from "react";
import { Link } from "react-router-dom";
import { dataTypes } from "constants/constants";
import { FormattedMessage } from "react-intl";
import { statusColor } from "functions/colors";
import moment from "moment";
import store from "../../../../store";
const { TEXT, PRICE, ACTIONS, FUNC, ASSIGNBOOKING, REFUNDBOOKING } = dataTypes;

/**
 * @name DisplayDateCityTime
 * @example of text output
 *     January 12, 2022
 *     Riyadh
 *     2:00 AM
 * @param {object} record
 * @param {string} locale
 * @param {string} location "DropOffCityName" || "PickUpCityName"
 * @param {string} date     "pickUpDate" || "dropOffDate"
 * @return {JSX}
 */

function DisplayDateCityTime(record, locale, location, date, time) {
  return (
    <div style={{ minWidth: "max-content", fontSize: "small" }}>
      {moment(record[date]).locale(locale).format("LL")}
      <br />
      <div style={{ maxWidth: "100px" }}>{record[`${locale}${location}`]}</div>
      {moment(record[time], "HH:mm:ss").locale(locale).format("hh:mm A")}
    </div>
  );
}
const { ally_id } = store.getState()?.authUser.user;

export const bookingTableData = [
  {
    headerId: "bookingId.placeholder",
    dataType: TEXT,
    dataRef: "id",
    orderBy: "id",
    issortable: true,
  },
  {
    headerId: "bookings.list.customerName",
    dataRef: "customerName",
    dataType: FUNC,
    func: (record) => (
      <Link to={`/cw/dashboard/customers/${record.userId}`}>{record.customerName}</Link>
    ),
  },
  {
    headerId: "bookings.list.allyName",
    dataRef: "AllyName",
    dataType: FUNC,
    func: (record, locale) => (
      <Link to={`/cw/dashboard/companies/${record.allyCompanyId}`}>
        {record[`${locale}AllyName`]}
      </Link>
    ),
  },
  { headerId: "branch", dataRef: "branchName", dataType: TEXT },

  {
    headerId: "car",
    bilingual: { ar: "arMakeName", en: "enMakeName" },
    dataType: FUNC,
    func: (record, locale) =>
      `${record[`${locale}MakeName`]} - ${record[`${locale}ModelName`]} - 
      ${record[`${locale}VersionName`]} - ${record.year}`,
  },
  {
    headerId: "bookings.list.numberOfDaysToBeRented",
    dataRef: "numberOfDays",
    dataType: TEXT,
    orderBy: "number_of_days",
    issortable: true,
  },
  { headerId: "paymentMethod", dataType: REFUNDBOOKING },

  {
    headerId: "bookings.list.carRentPricePerDay",
    dataRef: "pricePerDay",
    dataType: PRICE,
    orderBy: "price_per_day",
    issortable: true,
  },
  {
    headerId: "bookings.list.header.billingAmount",
    dataRef: "totalBookingPrice",
    dataType: PRICE,
    orderBy: "total_booking_price",
    issortable: true,
  },
  { headerId: "bookings.list.paidAmount", dataRef: "totalInsurancePrice", dataType: PRICE },
  {
    headerId: "businessBookings.list.pickup_date_time",
    dataRef: "pickUpDate",
    dataType: FUNC,
    func: (record, locale) =>
      DisplayDateCityTime(record, locale, "PickUpCityName", "pickUpDate", "pickUpTime"),
    orderBy: "pick_up_datetime",
    issortable: true,
  },
  {
    headerId: "businessBookings.list.dropoff_date_time",
    dataRef: "dropOffDate",
    dataType: FUNC,
    func: (record, locale) =>
      DisplayDateCityTime(record, locale, "DropOffCityName", "dropOffDate", "dropOffTime"),
    orderBy: "drop_off_datetime",
    issortable: true,
  },
  {
    headerId: "bookings.list.bookingStatus",
    dataRef: "status",
    dataType: FUNC,
    func: (record) => (
      <>
        <div className="d-flex">
          {record.isIntegratedRental && <div className="badge badge-info mr-2">Api</div>}
          <div
            className={`badge badge-${statusColor(record.status)}`}
            style={{ backgroundColor: statusColor(record.status), maxWidth: "fit-content" }}
          >
            <p className="m-0 p-0 font-weight-light" style={{ fontSize: "small" }}>
              {record.status && <FormattedMessage id={record.status.toUpperCase()} />}
            </p>
            <p className="m-0 p-0 font-weight-light" style={{ fontSize: "small" }}>
              {record?.subStatus && <FormattedMessage id={record?.subStatus?.toUpperCase()} />}
            </p>
          </div>
        </div>
        <div>{record.lastRentalDateExtensionRequest && <FormattedMessage id="extend" />}</div>
      </>
    ),
  },
  {
    headerId: "createdAt",
    dataRef: "createdAt",
    dataType: FUNC,
    func: (record, locale) => {
      if (record?.createdAt) {
        const date = moment(record.createdAt).locale(locale);
        return (
          <>
            <span>{date.format("LL")}</span> <br /> <span>{date.format("hh:mm A")}</span>
          </>
        );
      }
      return "";
    },
    orderBy: "created_at",
    issortable: true,
  },
  !ally_id ? { headerId: "Assign", dataType: ASSIGNBOOKING } : null,
  { headerId: "common.actions", dataType: ACTIONS },
];
