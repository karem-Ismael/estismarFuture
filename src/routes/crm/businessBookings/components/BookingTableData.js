/* eslint-disable prettier/prettier */
import React from "react";
import { Link } from "react-router-dom";
import { dataTypes } from "constants/constants";
import { FormattedDate, FormattedMessage, FormattedTime } from "react-intl";
import { statusColor } from "functions/colors";
import moment from "moment";
import store from "../../../../store";
import Refund from "./Refund";
const { TEXT, PRICE, BILINGUAL, ACTIONS, FUNC, ASSIGNBOOKING, REFUNDBOOKING } = dataTypes;

/**
 * @name DisplayDateCityTime
 * @example of text output
 *     January 12, 2022
 *     Riyadh
 *     2:00 AM
 * @param {object} record
 * @param {string} locale
 * @param {string} location "DropOffCityName" || "PickUpCityName"
 * @param {string} date     "pickUpDate" || "dropOffDate"
 * @return {JSX}
 */
// function tConvert(time) {
//   // Check correct time format and split into components
//   time = time.toString().match(/^([01]\d|2[0-3])(:)([0-5]\d)(:[0-5]\d)?$/) || [time];

//   if (time.length > 1) {
//     // If time format correct
//     time = time.slice(1); // Remove full string match value
//     time[5] = +time[0] < 12 ? "AM" : "PM"; // Set AM/PM
//     time[0] = +time[0] % 12 || 12; // Adjust hours
//   }
//   return time.join(""); // return adjusted time or original string
// }
function DisplayDateCityTime(record, locale, location, dateTime) {
  return (
    <div style={{ minWidth: "max-content", fontSize: "small" }}>
      {moment(record[dateTime]).locale(locale).format("LL")}
      <br />
      <div style={{ maxWidth: "100px" }}>{record[`${locale}${location}`]}</div>
      {moment(record[dateTime]).locale(locale).format("hh:mm A")}
    </div>
  );
}
const { ally_id } = store.getState()?.authUser.user;

export const bookingTableData = [
  {
    headerId: "bookingId.placeholder",
    dataType: TEXT,
    dataRef: "id",
    orderBy: "id",
    issortable: true,
  },
  { headerId: "bookingNo.placeholder", dataType: TEXT, dataRef: "bookingNo" },
  {
    headerId: "bookings.list.customerName",
    dataRef: "customerName",
    dataType: FUNC,
    func: (record) => (
      <Link to={`/cw/dashboard/customers/${record.userId}`}>{record.customerName}</Link>
    ),
  },
  {
    headerId: "bookings.list.allyName",
    dataType: FUNC,
    func: (record, locale) => (
      <Link to={`/cw/dashboard/companies/${record?.acceptedOffer?.allyCompanyId}`}>
        {record?.acceptedOffer?.allyCompanyName}
      </Link>
    ),
  },
  { headerId: "rental.insuranceType", dataRef: "insuranceName", dataType: TEXT },
  { headerId: "bookings.list.pickupCity", dataRef: "pickUpCityName", dataType: TEXT },

  {
    headerId: "car",
    bilingual: { ar: "arMakeName", en: "enMakeName" },
    dataType: FUNC,
    func: (record, locale) =>
      `${record[`${locale}MakeName`]} - ${record[`${locale}ModelName`]} - ${record.year}`,
  },
  {
    headerId: "Duration in months",
    dataRef: "numberOfMonths",
    dataType: TEXT,
    orderBy: "number_of_months",
    issortable: true,
  },
  { headerId: "numberOfCars", dataRef: "numberOfCars", dataType: TEXT },
  {
    headerId: "bookings.list.pickup",
    dataRef: "pickUpDatetime",
    dataType: FUNC,
    func: (record, locale) =>
      DisplayDateCityTime(record, locale, "PickUpCityName", "pickUpDateTime"),
    orderBy: "pick_up_datetime",
    issortable: true,
  },
  {
    headerId: "bookings.list.delivery",
    dataRef: "dropOffDatetime",
    dataType: FUNC,
    func: (record, locale) =>
      DisplayDateCityTime(record, locale, "DropOffCityName", "dropOffDateTime"),
    orderBy: "drop_off_datetime",
    issortable: true,
  },
  {
    headerId: "bookings.list.header.billingAmountMonth",
    dataRef: "pricePerMonth",
    dataType: TEXT,
    orderBy: "price_per_month",
    issortable: true,
  },
  {
    headerId: "bookings.list.header.billingAmount",
    dataRef: "totalBookingPrice",
    dataType: TEXT,
    orderBy: "total_booking_price",
    issortable: true,
  },
  {
    headerId: "bookings.list.paidAmount",
    func: (record) =>
      `${
        record?.acceptedOffer?.carInsuranceFull * record?.numberOfMonths ||
        record?.acceptedOffer?.carInsuranceStandard * record?.numberOfMonths
      }`,
    dataType: FUNC,
  },
  // {
  //   headerId: "businessBookings.list.dropoff_date_time",
  //   dataRef: "dropOffDatetime",
  //   dataType: FUNC,
  //   func: (record) => {
  //     const date = moment.utc(record.dropOffDatetime);
  //     return date.local().format("DD/MM/YYYY HH:mm:ss");
  //   },
  // },
  {
    headerId: "bookings.list.bookingStatus",
    dataRef: "status",
    dataType: FUNC,
    func: (record) => (
      <div className="d-flex">
        {record.isIntegratedRental && <div className="badge badge-info mr-2">Api</div>}
        <div
          className={`badge badge-${statusColor(record.status)}`}
          style={{ backgroundColor: statusColor(record.status), maxWidth: "fit-content" }}
        >
          <p className="m-0 p-0 font-weight-light" style={{ fontSize: "small" }}>
            {record.status && <FormattedMessage id={record.status.toUpperCase()} />}
          </p>
          <p className="m-0 p-0 font-weight-light" style={{ fontSize: "small" }}>
            {record?.subStatus && <FormattedMessage id={record?.subStatus?.toUpperCase()} />}
          </p>
        </div>
      </div>
    ),
  },
  {
    headerId: "createdAt",
    dataRef: "createdAt",
    dataType: FUNC,
    func: (record, locale) => {
      if (record?.createdAt) {
        const date = moment(record.createdAt).locale(locale);
        return (
          <>
            <span>{date.format("LL")}</span> <br /> <span>{date.format("hh:mm A")}</span>
          </>
        );
      }
      return "";
    },
    orderBy: "created_at",
    issortable: true,
  },
  !ally_id ? { headerId: "Assign", dataType: ASSIGNBOOKING } : null,
  { headerId: "common.actions", dataType: ACTIONS },
];
