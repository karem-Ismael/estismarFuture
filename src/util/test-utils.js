// test-utils.js
import React from "react";
import ReactTestUtils from "react-dom/test-utils";
import { render as rtlRender } from "@testing-library/react";
import { IntlProvider } from "react-intl";
import { translationMessages } from "../translations/i18n";

function render(ui, { locale = "en", ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return (
      <IntlProvider locale={locale} messages={translationMessages[locale]}>
        {children}
      </IntlProvider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

function renderWithIntl(element, locale = "en") {
  let instance;

  ReactTestUtils.renderIntoDocument(
    <IntlProvider locale={locale} messages={translationMessages[locale]}>
      {React.cloneElement(element, {
        ref: instance,
      })}
    </IntlProvider>,
  );

  return instance;
}
// re-export everything
export * from "@testing-library/react";

// override render method
export { render, renderWithIntl };
