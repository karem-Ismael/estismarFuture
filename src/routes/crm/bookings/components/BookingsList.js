/* eslint-disable prettier/prettier */
/* eslint-disable react/prop-types */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable eqeqeq */
/* eslint-disable no-nested-ternary */
/** Bookings List */
import React, { useEffect, useState } from "react";
import { FormattedMessage, useIntl } from "react-intl";
import { Link, useHistory } from "react-router-dom";
import PropTypes from "prop-types";
import { Typography, Tooltip, Button } from "@material-ui/core";
import { Pagination } from "@material-ui/lab";

// import useSetState from "hooks/useSetState";
import RctCollapsibleCard from "components/RctCollapsibleCard/RctCollapsibleCard";
import CustomTable from "components/shared/CustomTable";
import PerPage from "components/shared/PerPage";
import { userCan } from "functions/userCan";
import { daysDifference } from "functions";

import { useQuery, useMutation } from "@apollo/client";
import { AssignRentalTo } from "gql/mutations/AssignRentalTo.gql";
import { AssignRentalToMe } from "gql/mutations/AssignBookingToMe.gql";
import { PrintRental } from "gql/mutations/PrintRental.gql";
import NotificationManager from "react-notifications/lib/NotificationManager";
import swal from "sweetalert";
import TimeLine from "./TimeLineModal";
import store from "../../../../store";
import { bookingTableData } from "./BookingTableData";
import UsersModal from "./UsersModal";
import Refund from "./Refund";
import print from "print-js";
import ExtendModal from "./ExtendModal";
import CircularProgress from '@material-ui/core/CircularProgress';

const { is_super_user, ally_id } = store.getState()?.authUser.user;

function BookingsList({
  bookingsRes,
  loading,
  setPage,
  limit,
  setLimit,
  users,
  refetch,
  setOrderBy,
  setSortBy,
  decodedUserId,
}) {
  const history = useHistory();
  const { formatMessage } = useIntl();
  const [OpenUsersModal, setOpenUsersModal] = useState(false);
  const [bookingsState, setBookingsState] = useState({
    collection: [],
    metadata: {},
  });
  const [assignRentalTo] = useMutation(AssignRentalTo);
  const [printrental] = useMutation(PrintRental);
  const [assignRentalToMe] = useMutation(AssignRentalToMe);
  const { collection, metadata } = bookingsState;
  const [customerCare, setCustomerCare] = useState([]);
  const [BookingRecord, setBookingRecord] = useState({});
  const [opneTimeLineModal, setOpenTimeLineModal] = useState(false);
  const [BookingId, setBookingId] = useState();
  const [openmodel, setOpenModel] = useState(false);
  const[printProgress,setPrintProgress]=useState(false)

  useEffect(() => {
    setBookingsState({
      collection: bookingsRes?.dashboardRentals?.collection,
      metadata: bookingsRes?.dashboardRentals?.metadata,
    });
  }, [bookingsRes]);
  useEffect(() => {
    if (users?.length) {
      setCustomerCare(users);
    }
  }, [users]);
  const actions = ({ id, dropOffDate, pickUpDate, status,lastRentalDateExtensionRequest }) => {
    const [DOdate, PUdate] = [dropOffDate, pickUpDate];

    const getRentalAudits = (id) => {
      setBookingId(id);
      setOpenTimeLineModal(true);
    };
    const handelPrint = (id) => {
      setPrintProgress(true)
      printrental({
        variables: {
          rentalId: id,
        },
      }).then((data) => {
        setPrintProgress(false)
        print({ printable: data.data.printRental.fileBase64, type: "pdf", base64: true });
      });
    };
    const getRentalExtend = (id) => {
      setBookingId(id);
      setOpenModel(true);
    };
    return (
      <div className="d-flex align-items-center" style={{ gap: "5px" }}>
        {/* Redirects to booking/rental details */}
        <Tooltip title={formatMessage({ id: "common.details" })} placement="top">
          <Link to={`bookings/${id}`}>
            <i className="fa fa-eye"></i>
          </Link>
        </Tooltip>
        {userCan("rentals.update") && (
          <>
            {/* CAN EXTEND */}
            {daysDifference(PUdate, DOdate, status) === "IN_BETWEEN" && (
              <>
                <Tooltip title={formatMessage({ id: "extend" })} placement="top">
                  <Link to={`bookings/${id}/extend`}>
                    <i className="fa fa-calendar-plus-o"></i>
                  </Link>
                </Tooltip>
              </>
            )}
            {/* CAN CHANGE STATUS */}
            {daysDifference(PUdate, DOdate, status) !== "STOP_EDIT" && (
              <Tooltip title={formatMessage({ id: "changeStatus" })} placement="top">
                <Link to={`changestatus/${id}`}>
                  <i className="fa fa-refresh"></i>
                </Link>
              </Tooltip>
            )}
            {/* CAN EDIT */}
            {/* THIS Part of code is the correct one and it will be applied after an approval from the client */}
            {daysDifference(PUdate, DOdate, status) !== "STOP_EDIT" && status !== "car_received" && (
              // <Tooltip title={formatMessage({ id: "common.edit" })} placement="top">
              //   <Link to={`bookings/${id}/edit`}>
              //     <i className="fa fa-edit mr-1 ml-1"></i>
              //   </Link>
              // </Tooltip>
              <></>
            )}
            <Tooltip title={formatMessage({ id: "common.edit" })} placement="top">
              <Link to={`bookings/${id}/edit`}>
                <i className="fa fa-edit"></i>
              </Link>
            </Tooltip>
            {
              userCan("rentals.print") && 
              <Tooltip title={formatMessage({ id: "common.print" })} placement="top">
              <i
                className="fa fa-print"
                style={{ color: "#5d89d8",cursor:"pointer" }}
                onClick={() => handelPrint(id)}
              ></i>
              {/* karem */}
            </Tooltip>
            }
           

            {is_super_user && (
              <Tooltip title={formatMessage({ id: "common.timeline" })} placement="top">
                <Link>
                  <i className="fas fa-history" onClick={() => getRentalAudits(id)}></i>
                </Link>
              </Tooltip>
            )}
          </>
        )}
         {lastRentalDateExtensionRequest && userCan("rentals.extend")&& (
          <Tooltip title={formatMessage({ id: "extend" })}>
            <i
              className="fa fa-external-link"
              onClick={() => getRentalExtend(id)}
              style={{ color: "#5d89d8", cursor: "pointer" }}
            ></i>
            
          </Tooltip>
        )}

        {/* <div>
          <ExtendModal setOpenModel={setOpenModel} openmodel={openmodel} />
        </div> */}
      </div>
    );
  };

  const AssignUserToBooking = (record, isSuperAdmin) => {
    if (isSuperAdmin) {
      setBookingRecord(record);
      setOpenUsersModal(true);
    } else {
      swal({
        title: formatMessage({ id: "are.u.sure.?" }),
        text: formatMessage({ id: `u.want.to.Assign.this.Booking.to.u` }),
        icon: "warning",
        buttons: [formatMessage({ id: "cancel" }), formatMessage({ id: "button.yes" })],
        dangerMode: true,
      }).then((Yes) => {
        if (Yes) {
          assignRentalToMe({
            variables: {
              rentalId: +record.id,
            },
          }).then(() => {
            refetch();
            NotificationManager.success(<FormattedMessage id="BookingAssignedSuccessfully" />);
          });
        }
      });
    }
  };
  const AssignBookingBySuperUser = (BookingDetails, customerid) => {
    const customerName = customerCare.find((customer) => +customer.id == +customerid)?.name;
    if (!customerid) {
      NotificationManager.error(<FormattedMessage id="please.select.one.of.customercare" />);
      return;
    }
    swal({
      title: formatMessage({ id: "are.u.sure.?" }),
      text: `${formatMessage({ id: `u.want.to.Assign.this.Booking` })} ${customerName}`,
      icon: "warning",
      buttons: [formatMessage({ id: "cancel" }), formatMessage({ id: "button.yes" })],
      dangerMode: true,
    }).then((Yes) => {
      if (Yes) {
        assignRentalTo({
          variables: {
            rentalId: BookingDetails.id,
            userId: customerid,
          },
        }).then(() => {
          refetch();
          NotificationManager.success(<FormattedMessage id="BookingAssignedSuccessfully" />);
          setOpenUsersModal(false);
        });
      }
    });
  };
  const AssignBooking = (record) => (
    <>
      {record.assignedTo && !is_super_user ? null : (
        <Button
          variant="contained"
          color="primary"
          className="mx-smt-15 btn btn-danger mr-1 ml-1"
          onClick={() => {
            AssignUserToBooking(record, is_super_user);
          }}
        >
          <FormattedMessage id={!is_super_user ? "AssignToMe" : "AssignTo"} />
        </Button>
      )}

      {customerCare.find((customer) => +customer.id == +record.assignedTo)?.name}
    </>
  );
  const refund = (record) =>
    userCan("rentals.refund") && record.refundable ? (
      <div style={{ textAlign: "center" }}>
        <p>{record.paymentMethod}</p>
        {record.isPaid && record.paymentMethod == "ONLINE" ? (
          <>
            <p className="badge badge-info">
              {" "}
              <FormattedMessage id="payed" />
            </p>
          </>
        ) : !record.isPaid &&
          record.paymentMethod == "ONLINE" &&
          !record.paymentMethod == "CASH" ? (
          <FormattedMessage id="notpayed" />
        ) : null}
        {!ally_id && (
          <Refund
            rentalid={record.id}
            is24Passed={record.is24Passed}
            refetch={refetch}
            bookingDetails={record}
          />
        )}
      </div>
    ) : (
      <div>
        <p>
          {" "}
          <FormattedMessage id={record?.paymentMethod} />{" "}
        </p>
        {record.isPaid && record.paymentMethod == "ONLINE" ? (
          <>
            {!record.refundedAt && (
              <p className="badge badge-success">
                <FormattedMessage id="payed" />
              </p>
            )}
            {record.refundedAt && (
              <p className="badge badge-danger">
                <FormattedMessage id="refund.money" />
              </p>
            )}
          </>
        ) : !record.isPaid && record.paymentMethod == "ONLINE" ? (
          <FormattedMessage id="notpayed" />
        ) : null}
      </div>
    );
  return (
    <Typography component="div" style={{ padding: 10,position:"relative" }}>
      <div>
      {/* {
          printProgress && 
            <div className="progress-print" style={{width:"calc(100vw - 8.1rem)",height:"100vh",position:"absolite", display: "flex", justifyContent: "center", alignItems:"center", zIndex:"99", top:"45vh", left:"0"}}>
            <CircularProgress/>

            </div>

          } */}
        <UsersModal
          isOpen={OpenUsersModal}
          setOpenUsersModal={setOpenUsersModal}
          customerCare={customerCare}
          BookingDetails={BookingRecord}
          AssignBookingBySuperUser={AssignBookingBySuperUser}
        />
        <TimeLine
          isOpen={opneTimeLineModal}
          setOpenTimeLineModal={setOpenTimeLineModal}
          BookingId={BookingId}
        />
         <ExtendModal
          setOpenModel={setOpenModel}
          openmodel={openmodel}
          BookingId={BookingId}
          refetch={refetch}
        />
        <RctCollapsibleCard fullBlock table>
          
          <CustomTable
            tableData={bookingTableData}
            loading={printProgress || loading}
            tableRecords={collection}
            actions={actions}
            AssignBooking={AssignBooking}
            RefundBooking={refund}
            actionsArgs={["id", "pickUpDate", "dropOffDate", "status", "pickUpTime", "dropOffTime","lastRentalDateExtensionRequest"]}
            setOrderBy={setOrderBy}
            setSortBy={setSortBy}
          />
         
        </RctCollapsibleCard>
      </div>
      <div className="d-flex justify-content-around">
        {metadata?.currentPage && (
          <>
            <Pagination
              showFirstButton
              showLastButton
              count={Math.ceil(metadata?.totalCount / limit)}
              page={metadata?.currentPage}
              onChange={(e, value) => {
                setPage(value);
                history.replace({ hash: `page=${value}`, search: history?.location?.search });
              }}
            />
            <PerPage
              handlePerPageChange={(value) => {
                setLimit(value);
                setPage(1);
                history.replace({ hash: `page=1` });
              }}
              perPage={limit}
              setPage={setPage}
            />
          </>
        )}
      </div>
    </Typography>
  );
}

BookingsList.propTypes = {
  setPage: PropTypes.func,
  setLimit: PropTypes.func,
  loading: PropTypes.bool,
  bookingsRes: PropTypes.object,
  limit: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};

export default BookingsList;
