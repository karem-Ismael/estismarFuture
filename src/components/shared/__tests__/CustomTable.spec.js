import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "util/test-utils";
import { bookingTableData } from "routes/crm/bookings/components/BookingTableData";
import CustomTable from "../CustomTable";

const records = [
  {
    id: "69",
    totalBookingPrice: 148,
    totalInsurancePrice: 112,
    pickUpDate: "2020-12-15",
    dropOffDate: "2020-12-18",
    enMakeName: "Buick",
    arMakeName: "Dodge",
    arModelName: "Juke",
    enModelName: "Prius",
    year: "2016",
    enPickUpCityName: "Riyadh",
    arPickUpCityName: "الرياض",
    enDropOffCityName: "Riyadh",
    arDropOffCityName: "الرياض",
    enVersionName: "ESi",
    arVersionName: "L",
    status: "NEW",
  },
  {
    id: "71",
    totalBookingPrice: 92,
    totalInsurancePrice: 272,
    pickUpDate: "2020-12-15",
    dropOffDate: "2020-12-18",
    enMakeName: "Ford",
    arMakeName: "Honda",
    arModelName: "M5",
    enModelName: "Regal",
    year: "2019",
    enPickUpCityName: "Riyadh",
    arPickUpCityName: "الرياض",
    enDropOffCityName: "Riyadh",
    arDropOffCityName: "الرياض",
    enVersionName: "XL",
    arVersionName: "ESi",
    status: "NEW",
  },
];

test("it should render FormattedDate and have a formated pt date", () => {
  render(
    <CustomTable
      tableData={bookingTableData}
      tableRecords={records}
      actions={(id) => <div>Actions {id}</div>}
    />,
    { locale: "en" },
  );
  expect(screen.getByTestId("header-row")).toHaveTextContent("Actions");
  expect(screen.getByTestId("data-tr-0")).toHaveTextContent("Buick - Prius - ESi - 2016");
  // to ensure that the booking table data didn't change.
  // if it changed we should update the following test
  expect(bookingTableData).toMatchSnapshot();
  expect(screen.getByTestId("data-tr-0")).toHaveTextContent(
    /Buick - Prius - ESi - 2016 SAR148 SAR112 SARDecember 15, 2020Riyadh2:00 AMDecember 18, 2020Riyadh2:00 AMNew/i,
  );
});

it("should capitcalize all heads", () => {
  render(<CustomTable tableData={["abc"]} />);
  screen.getByTestId("capitalized-header");
});
