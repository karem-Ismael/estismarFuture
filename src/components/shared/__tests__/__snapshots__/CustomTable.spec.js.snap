// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render FormattedDate and have a formated pt date 1`] = `
Array [
  Object {
    "dataRef": "bookingNo",
    "dataType": "TEXT",
    "headerId": "bookingNo.placeholder",
  },
  Object {
    "dataRef": "customerName",
    "dataType": "TEXT",
    "headerId": "bookings.list.customerName",
  },
  Object {
    "bilingual": Object {
      "ar": "arAllyName",
      "en": "enAllyName",
    },
    "dataType": "BILINGUAL",
    "headerId": "bookings.list.allyName",
  },
  Object {
    "bilingual": Object {
      "ar": "arMakeName",
      "en": "enMakeName",
    },
    "dataType": "FUNC",
    "func": [Function],
    "headerId": "car",
  },
  Object {
    "dataRef": "numberOfDays",
    "dataType": "TEXT",
    "headerId": "bookings.list.numberOfDaysToBeRented",
  },
  Object {
    "dataRef": "pricePerDay",
    "dataType": "PRICE",
    "headerId": "bookings.list.carRentPricePerDay",
  },
  Object {
    "dataRef": "totalBookingPrice",
    "dataType": "PRICE",
    "headerId": "bookings.list.billingAmount",
  },
  Object {
    "dataRef": "totalInsurancePrice",
    "dataType": "PRICE",
    "headerId": "bookings.list.paidAmount",
  },
  Object {
    "dataRef": "pickUpDate",
    "dataType": "FUNC",
    "func": [Function],
    "headerId": "bookings.list.pickup",
  },
  Object {
    "dataRef": "pickUpDate",
    "dataType": "FUNC",
    "func": [Function],
    "headerId": "bookings.list.delivery",
  },
  Object {
    "dataRef": "status",
    "dataType": "FUNC",
    "func": [Function],
    "headerId": "bookings.list.bookingStatus",
  },
  Object {
    "dataType": "ACTIONS",
    "headerId": "common.actions",
  },
]
`;
